packages/costa/src/runtime.ts

```diff
@@ -190,31 +202,42 @@ export class CostaRuntime {
   }
   /**
    * fetch and check if the package name is valid.
-   * @param name the plugin package name.
+   * @param name the plugin package name or ReadStream for npm package tarball.
    * @param cwd the current working directory.
    */
-  async fetch(name: string, cwd?: string): Promise<PluginPackage> {
-    const source = this.getSource(name, cwd || process.cwd());
+  async fetch(name: string | Readable, cwd?: string): Promise<PluginPackage> {
     let pkg: PluginPackage;
-    if (source.from === 'npm') {
-      debug(`requesting the url ${source.uri}`);
-      // TODO(yorkie): support http cache
-      const resp = await requestHttpGetWithCache(source.uri);
-      const meta = resp as NpmPackageMetadata;
-      pkg = selectNpmPackage(meta, source);
-    } else if (source.from === 'git') {
-      debug(`requesting the url ${source.uri}...`);
-      const { hostname, auth, hash } = source.urlObject;
-      let pathname = source.urlObject.pathname.replace(/^\/:?/, '');
-      const remote = `${auth || 'git'}@${hostname}:${pathname}${hash || ''}`;
-      pkg = await fetchPackageJsonFromGit(remote, 'HEAD');
-    } else if (source.from === 'fs') {
-      debug(`linking the url ${source.uri}`);
-      pkg = require(`${source.uri}/package.json`);
-    } else if (source.from === 'tarball') {
-      debug(`downloading the url ${source.uri}`);
-      await download(source.name, source.uri);
-      pkg = await fetchPackageJsonFromTarball(source.uri);
+    let source: PluginSource;
+    if (typeof name === 'string') {
+      source = this.getSource(name, cwd || process.cwd());
+      if (source.from === 'npm') {
+        debug(`requesting the url ${source.uri}`);
+        // TODO(yorkie): support http cache
+        const resp = await requestHttpGetWithCache(source.uri);
+        const meta = resp as NpmPackageMetadata;
+        pkg = selectNpmPackage(meta, source);
+      } else if (source.from === 'git') {
+        debug(`requesting the url ${source.uri}...`);
+        const { hostname, auth, hash } = source.urlObject;
+        let pathname = source.urlObject.pathname.replace(/^\/:?/, '');
+        const remote = `${auth || 'git'}@${hostname}:${pathname}${hash || ''}`;
+        pkg = await fetchPackageJsonFromGit(remote, 'HEAD');
+      } else if (source.from === 'fs') {
+        debug(`linking the url ${source.uri}`);
+        pkg = require(`${source.uri}/package.json`);
+      } else if (source.from === 'tarball') {
+        debug(`downloading the url ${source.uri}`);
+        await download(source.name, source.uri);
+        pkg = await fetchPackageJsonFromTarball(source.uri);
+      }
+    } else {
+      source = {
+        from: 'tarball',
+        uri: null,
+        urlObject: undefined,
```

Comment:
Maybe we could mark the above `uri` and `urlObject` to be optional so that we could not set it here.